File: theoTracing.m
function [B,L] = theoTracing(region)
B = {};
L = {};

traceRegion = region; 
marker = zeros(size(traceRegion));

disp(traceRegion(600:620,1980:2000));
% disp(1);

[x,y] = size(region);
numcells = 0; % Current cell of B

for i = 1:x,
    for j = 1:y,

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        if(and(j + 1 < y,i + 1 < x)),
            entryLocation = []; 
            s = []; 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            if((traceRegion(i,j+1) == 0) && traceRegion(i,j) ~= traceRegion(i,j+1) && marker(i,j+1)~=-1), %white to black detection
                numcells = numcells + 1;
                numRows = 1;
                prev = [];
                entryLocation(1) = i;
                entryLocation(2) = j;
                s(1) = i;
                s(2) = j+1;
                prev(1) = i;
                prev(2) = j;
                currentPx = s(1);
                currentPy = s(2);
                B{numcells}(numRows,1) = currentPx;
                B{numcells}(numRows,2) = currentPy;
                marker(currentPx,currentPy) = -1;
                
                if numcells == 10,
                    disp(currentPx);
                    disp(currentPy);
                    disp(1);
                end
                
                numRows = numRows + 1;
                [currentPx,currentPy,prev] = getTheoNeighbourhood(traceRegion,currentPx,currentPy,prev);
                B{numcells}(numRows,1) = currentPx;
                B{numcells}(numRows,2) = currentPy;
                marker(currentPx,currentPy) = -1;
                
                if numcells == 10,
                    n = 0;
                    while n < 20,
                        numRows = numRows + 1;
                        [currentPx,currentPy,prev] = getTheoNeighbourhood(traceRegion,currentPx,currentPy,prev);
                        B{numcells}(numRows,1) = currentPx;
                        B{numcells}(numRows,2) = currentPy;
                        marker(currentPx,currentPy) = -1;
                        n = n + 1;
                    end
                end
                
                if numcells == 10,
                    disp(B{10});
                    imshow(region);
                    hold on
                    boundary = B{10};
                    plot(boundary(:,2), boundary(:,1), 'r', 'LineWidth', 2);
                    disp(1);
                end
                
                n = 0;
                while n < 12000, %(s(1) ~= currentPx) | (s(2)~=currentPy),
                    numRows = numRows + 1;
                    [currentPx,currentPy,prev] = getTheoNeighbourhood(traceRegion,currentPx,currentPy,prev);
                    B{numcells}(numRows,1) = currentPx;
                    B{numcells}(numRows,2) = currentPy;
                    marker(currentPx,currentPy) = -1;
                    n = n + 1;
                end
            end
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        end
    end
end

end

File: python/domino_helper/theo_tracing.py
"""
theoTracing.m -> theo_tracing.py

Alternative boundary tracing implementation (debug/experimental version).
Uses get_theo_neighbourhood instead of get_neighbourhood.
"""

import numpy as np
from .get_theo_neighbourhood import get_theo_neighbourhood


def theo_tracing(region):
    """
    Trace boundaries using the alternative theo neighbourhood method.

    Parameters
    ----------
    region : np.ndarray
        Binary image (values 0 and 1).

    Returns
    -------
    B : list of np.ndarray
        List of boundary arrays, each of shape (N, 2) with [row, col] coordinates.
    L : list
        Placeholder.
    """
    B = []
    L = []

    trace_region = region.copy()
    x, y = region.shape
    marker = np.zeros((x, y), dtype=np.int32)

    numcells = 0

    for i in range(x):
        for j in range(y):
            if j + 1 < y and i + 1 < x:
                if (trace_region[i, j + 1] == 0 and
                        trace_region[i, j] != trace_region[i, j + 1] and
                        marker[i, j + 1] != -1):

                    numcells += 1
                    boundary = []

                    s = [i, j + 1]
                    prev = [i, j]
                    current_px = s[0]
                    current_py = s[1]

                    boundary.append([current_px, current_py])
                    marker[current_px, current_py] = -1

                    current_px, current_py, prev = get_theo_neighbourhood(
                        trace_region, current_px, current_py, prev
                    )
                    boundary.append([current_px, current_py])
                    marker[current_px, current_py] = -1

                    # Fixed iteration count (matching MATLAB's n < 12000)
                    for _ in range(12000):
                        current_px, current_py, prev = get_theo_neighbourhood(
                            trace_region, current_px, current_py, prev
                        )
                        boundary.append([current_px, current_py])
                        marker[current_px, current_py] = -1

                    B.append(np.array(boundary))

    return B, L
